pkg/controller/stack/stack_controller.go

```diff
@@ -599,7 +599,22 @@ func (sess *reconcileStackSession) SetupPulumiWorkdir(gitAuth *auto.GitAuth) err
 	sess.logger.Debug("Setting up pulumi workdir for stack", "stack", sess.stack)
 	// Create a new workspace.
 	secretsProvider := auto.SecretsProvider(sess.stack.SecretsProvider)
-	w, err := auto.NewLocalWorkspace(context.Background(), auto.Repo(repo), secretsProvider)
+
+	// Create the temporary workdir
+	dir, err := os.MkdirTemp("", "pulumi_auto")
+	if err != nil {
+		return errors.Wrap(err, "unable to create tmp directory for workspace")
+	}
+
+	// Cleanup the workdir on failure setting up the workspace.
+	defer func() {
+		if err != nil {
+			_ = os.RemoveAll(dir)
+		}
+	}()
+
+	var w auto.Workspace
+	w, err = auto.NewLocalWorkspace(context.Background(), auto.WorkDir(dir), auto.Repo(repo), secretsProvider)
```

Comment:
What's the default behavior of our automation API before this change? Does it create a temporary folder internally but does not clean it up correctly? What is our advice to 3rd party users of automation API?

Reply:
> What's the default behavior of our automation API before this change? Does it create a temporary folder internally but does not clean it up correctly? What is our advice to 3rd party users of automation API?

Yes it creates a temporary directory internally. We tried to handle cleanup through the `sess.CleanupPulumiWorkdir()` call elsewhere but we do many operations as part of the stack setup which may fail independently and we leak the work dir on failure.
This seems to be more a function of the way the operator is currently written than anything specific to the automation API.